altair/utils/_vegafusion_data.py

```diff
@@ -64,7 +68,22 @@ def vegafusion_data_transformer(
     """VegaFusion Data Transformer."""
     if data is None:
         return vegafusion_data_transformer
-    elif isinstance(data, DataFrameLike) and not isinstance(data, SupportsGeoInterface):
+
+    # Test whether VegaFusion supports the data type
+    is_v2 = (
+        Version(version) >= Version("2.0.0a0")
+        if (version := importlib_version("vegafusion"))
+        else False
+    )
+    if is_v2:
+        # VegaFusion v2 support narwhals-compatible DataFrames
+        supported_by_vf = isinstance(data, DataFrameLike) or isinstance(
+            nw.from_native(data, pass_through=True), (nw.DataFrame, nw.LazyFrame)
+        )
+    else:
+        supported_by_vf = isinstance(data, DataFrameLike)
+
+    if supported_by_vf and not isinstance(data, SupportsGeoInterface):
```

Comment:
@jonmmease I haven't tested this, but would you be able to preserve the typing of `extracted_inline_tables` by rewriting `supported_by_vf` in this style?

https://github.com/dangotbanned/altair/blob/0c829fb438c50955128b747d96480e05a2dd4fb2/tools/schemapi/schemapi.py#L453-L469

https://github.com/dangotbanned/altair/blob/0c829fb438c50955128b747d96480e05a2dd4fb2/tools/schemapi/schemapi.py#L266-L450

---

There are a lot more examples of this kind of conditional defintion blocks in:
https://github.com/python/typing_extensions/blob/e888dfdabcfcd03330477cb5de3c25c21029f425/src/typing_extensions.py

---

I'm imagining something like:

```py

if (version := importlib_version("vegafusion")) and Version(version) >= Version("2.0.0a0"):
    def is_supported_by_vf(data: Any) -> TypeIs[DataFrameLike]:
        ...
else:
    def is_supported_by_vf(data: Any) -> TypeIs[DataFrameLike]:
        ...

```

Reply:
Thanks for the suggestion @dangotbanned. I played with this approach using `TypeIs` but wasn't successful in getting mypy to honor the narrowing.  But I realized the simpler approach is to keep using `DataFrameLike` and apply a `typing.cast` in VegaFusion v2 case ([a6853a9](https://github.com/vega/altair/pull/3682/commits/a6853a94500db237ec52964b122f2394972451ca)).

Does that work for you?

Reply:
@jonmmease 

Tested this locally and `mypy` seems to not complain:
```diff
diff --git a/altair/utils/_vegafusion_data.py b/altair/utils/_vegafusion_data.py
index 876cfc56..d08ea5ba 100644
--- a/altair/utils/_vegafusion_data.py
+++ b/altair/utils/_vegafusion_data.py
@@ -2,10 +2,10 @@ from __future__ import annotations
 
 import uuid
 from importlib.metadata import version as importlib_version
-from typing import TYPE_CHECKING, Any, Callable, Final, TypedDict, Union, cast, overload
+from typing import TYPE_CHECKING, Any, Callable, Final, TypedDict, Union, overload
 from weakref import WeakValueDictionary
 
-import narwhals.stable.v1 as nw
+from narwhals.dependencies import is_into_dataframe
 from packaging.version import Version
 
 from altair.utils._importers import import_vegafusion
@@ -19,12 +19,18 @@ from altair.utils.data import (
 from altair.vegalite.data import default_data_transformer
 
 if TYPE_CHECKING:
+    import sys
     from collections.abc import MutableMapping
 
     from narwhals.typing import IntoDataFrame
 
     from vegafusion.runtime import ChartState
 
+    if sys.version_info >= (3, 13):
+        from typing import TypeIs
+    else:
+        from typing_extensions import TypeIs
+
 # Temporary storage for dataframes that have been extracted
 # from charts by the vegafusion data transformer. Use a WeakValueDictionary
 # rather than a dict so that the Python interpreter is free to garbage
@@ -69,23 +75,9 @@ def vegafusion_data_transformer(
     if data is None:
         return vegafusion_data_transformer
 
-    # Test whether VegaFusion supports the data type
-    is_v2 = (
-        Version(version) >= Version("2.0.0a0")
-        if (version := importlib_version("vegafusion"))
-        else False
-    )
-    if is_v2:
-        # VegaFusion v2 support narwhals-compatible DataFrames
-        supported_by_vf = isinstance(data, DataFrameLike) or isinstance(
-            nw.from_native(data, pass_through=True), (nw.DataFrame, nw.LazyFrame)
-        )
-    else:
-        supported_by_vf = isinstance(data, DataFrameLike)
-
-    if supported_by_vf and not isinstance(data, SupportsGeoInterface):
+    if is_supported_by_vf(data) and not isinstance(data, SupportsGeoInterface):
         table_name = f"table_{uuid.uuid4()}".replace("-", "_")
-        extracted_inline_tables[table_name] = cast(DataFrameLike, data)
+        extracted_inline_tables[table_name] = data
         return {"url": VEGAFUSION_PREFIX + table_name}
     else:
         # Use default transformer for geo interface objects
@@ -94,6 +86,20 @@ def vegafusion_data_transformer(
         return default_data_transformer(data)
 
 
+if (version := importlib_version("vegafusion")) and Version(version) >= Version(
+    "2.0.0a0"
+):
+
+    def is_supported_by_vf(data: Any) -> TypeIs[DataFrameLike]:
+        # Test whether VegaFusion supports the data type
+        # VegaFusion v2 support narwhals-compatible DataFrames
+        return is_into_dataframe(data)
+else:
+
+    def is_supported_by_vf(data: Any) -> TypeIs[DataFrameLike]:
+        return isinstance(data, DataFrameLike)
+
+
 def get_inline_table_names(vega_spec: dict[str, Any]) -> set[str]:
     """
     Get a set of the inline datasets names in the provided Vega spec.

```

`is_into_dataframe` ends on a branch that is equivalent to `DataFrameLike`, so it should cover everything `narwhals` accepts

Reply:
Oh, nice. Yeah this does work. Not sure now what I was missing, but thanks!